substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,180 @@
+import org.graalvm.compiler.nodes.java.LoadFieldNode;
+import org.graalvm.compiler.nodes.java.StoreFieldNode;
+import org.graalvm.compiler.options.Option;
+import org.graalvm.compiler.options.OptionValues;
+import org.graalvm.compiler.phases.OptimisticOptimizations;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.oracle.graal.pointsto.infrastructure.GraphProvider.*;
+
+public class NativeImageInlineDuringParsingPlugin implements InlineInvokePlugin {
+
+    // to use this plugin: native-image Example -H:+InlineBeforeAnalysis
+    public static class Options {
+        @Option(help = "Inline methods during parsing before the static analysis.")//
+        public static final HostedOptionKey<Boolean> InlineBeforeAnalysis = new HostedOptionKey<>(false);
+
+    }
+
+    static class InvocationResult {
+        static final InvocationResult ANALYSIS_TOO_COMPLICATED = new InvocationResult();
+        static final InvocationResult NO_ANALYSIS = new InvocationResult();
+    }
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    @Override
+    public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod method, ValueNode[] args) {
+
+        if (b.parsingIntrinsic()) {
+            /* We are not interfering with any intrinsic method handling. */
+            return null;
+        }
+
+        if (method.getAnnotation(NeverInline.class) != null || method.getAnnotation(NeverInlineTrivial.class) != null) {
+            return null;
+        }
+
+        if (method.getAnnotation(RestrictHeapAccess.class) != null || method.getAnnotation(Uninterruptible.class) != null ||
+                b.getMethod().getAnnotation(RestrictHeapAccess.class) != null || b.getMethod().getAnnotation(Uninterruptible.class) != null) {
+            /*
+             * Caller or callee have an annotation that might prevent inlining. We don't check the
+             * exact condition but instead always bail out for simplicity.
+             */
+            return null;
+        }
+
+        if (!method.hasBytecodes())
+            return null;
+
+        if (method.isSynchronized())
+          return null;
+
+        if (providers.getGraphBuilderPlugins().getInvocationPlugins().lookupInvocation(method) != null)
+            return null;
+
+        if (((AnalysisMethod)method).buildGraph(b.getDebug(), method, providers, Purpose.ANALYSIS) != null)
+            /* Method has a manually constructed graph via GraphProvider. */
+            return null;
+
+        // method is declared in user's class
+        if(SubstrateOptions.Class.getValue().equals(method.format("%H"))) {
```

Comment:
@vjovanov 
izdvojene su za sada samo metode koje su korisnicki definisane, da se za njih dobije graf

Reply:
I would use the code from Christian here. He already went through the whole filtering process.